doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+
+- By blocking in `BIO_write` calls, this precludes correct implementation of
+  QUIC. Unlike any analogue in TLS, QUIC packets are time sensitive and intended
+  to be transmitted as soon as they are generated. QUIC packets contain fields
+  such as the ACK Delay value, which is intended to describe the time between a
+  packet being received and a return packet being generated. Correct calculation
+  of this field is necessary to correct calculation of connection RTT. It is
+  therefore important to only generate packets when they are ready to be sent,
+  otherwise suboptimal performance will result. This is a usage model which
+  aligns optimally to non-blocking I/O and which cannot be accommodated
+  by blocking I/O.
+
+- Since existing custom BIOs will not be expecting concurrent `BIO_read` and
+  `BIO_write` calls, they will need to be adapted to support this, which is
+  likely to require substantial rework of those custom BIOs (trivial locking of
+  calls obviously does not work since both of these calls must be able to block
+  on network I/O simultaneously).
+
+Moreover, this does not appear to be a realistically implementable approach:
+
+- The question is posed of how to handle connection teardown, which does not
+  seem to be solvable. If parallel threads are blocking in blocking `BIO_read`
+  and `BIO_write` calls on some underlying network BIO, there needs to be some
+  way to force these calls to return once `SSL_free` is called and we need to
+  tear down the connection. However, the BIO interface does not provide
+  any way to do this. *At best* we might assume the BIO is a `BIO_s_dgram`
+  (but cannot assume this in the general case), but even then we can only
+  accomplish teardown by violating the BIO abstraction and closing the
+  underlying socket.
+
+  This is the only portable way to ensure that a recv(3) call to the same socket
+  returns. This obviously is a highly application-visible change (and is likely
+  to be far more disruptive than configuring the socket into non-blocking mode).
+
+  Moreover, it is not workable anyway because it only works for a socket-based
+  BIO and violates the BIO abstraction. For BIOs in general, there does not
+  appear to be any viable solution to the teardown issue.
+
+Even if this approach were successfully implemented, applications will still
+need to change to using network BIOs with datagram semantics. For applications
+using custom BIOs, this is likely to require substantial rework of those BIOs.
+There is no possible way around this. Thus, even if this solution were adopted
+(notwithstanding the issues which preclude this noted above) for the purposes of
+accommodating applications using custom network BIOs in a blocking mode, these
+applications would still have to completely rework their implementation of those
+BIOs. In any case, it is expected to be very rare that sophisticated
+applications implementing their own custom BIOs will do so in a blocking mode.
```

Comment:
I'm not sure its necessarily that rare. I could definitely imagine some wrapper BIO that still uses sockets underneath but does it in its own way for whatever reason (e.g. because its using higher performing options/APIs that we don't use for portability reasons).